### PR TITLE
ARROW-8131: [Python] Add dynamic attributes to PyArrow ExtensionArray

### PR DESCRIPTION
**Pull request by balancap**

### CODE REVIEW

File: python/pyarrow/array.pxi [hunk 1]
```diff
@@ -2029,6 +2029,7 @@ cdef class ExtensionArray(Array):
     """
     Concrete class for Arrow extension arrays.
     """
+    cdef dict __dict__
```
Comment (kszucs) [inline python/pyarrow/array.pxi]: @balancap please add this to lib.pxd

**Comment by kszucs** [comment]
I'm not sure that we should support dynamic attributes, but please add tests cases which could reason about it.

cc @jorisvandenbossche

**Review by kszucs — CHANGES_REQUESTED** [review]
I'm not sure that we should support dynamic attributes, but please add tests cases which could reason about it.

**Comment by kszucs** [comment]
@balancap can we close this in favor of #6653?

**Comment by balancap** [comment]
Indeed. Closed the Jira ticket as well.

**Pull request closed without merging**